File: ios/ThirtySevenDegrees/DesignSystem/Molecules/PostCard.swift
import SwiftUI

struct PostCard: View {
    let avatarURL: URL?
    let username: String
    let handle: String
    let timeAgo: String
    let content: String
    let tags: [String]
    let imageURLs: [URL]
    let likeCount: Int
    let commentCount: Int
    let isLiked: Bool
    let onLike: () -> Void
    let onComment: () -> Void
    let onShare: () -> Void
    let onMore: () -> Void
    let onAvatarTap: () -> Void
    let onTagTap: (String) -> Void

    @State private var showLikeAnimation = false

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header
            if !content.isEmpty { contentSection }
            if !imageURLs.isEmpty { mediaSection }
            actionBar
        }
        .background(Color.bgPrimary)
    }

    private var header: some View {
        HStack(spacing: .spaceSM) {
            AvatarView(url: avatarURL, size: .lg, placeholder: username)
                .onTapGesture { onAvatarTap() }

            VStack(alignment: .leading, spacing: 2) {
                Text(username)
                    .font(.headline)
                    .foregroundStyle(.textPrimary)
                HStack(spacing: .spaceXS) {
                    Text("@\(handle)")
                        .font(.caption)
                        .foregroundStyle(.textTertiary)
                    Text("·")
                        .foregroundStyle(.textTertiary)
                    Text(timeAgo)
                        .font(.caption)
                        .foregroundStyle(.textTertiary)
                }
            }
            Spacer()
        }
        .padding(.horizontal, .spaceLG)
        .padding(.vertical, .spaceSM)
    }

    private var contentSection: some View {
        VStack(alignment: .leading, spacing: .spaceXS) {
            Text(content)
                .font(.bodyMedium)
                .foregroundStyle(.textPrimary)
                .lineLimit(6)

            if !tags.isEmpty {
                HStack(spacing: .spaceXS) {
                    ForEach(tags, id: \.self) { tag in
                        Text("#\(tag)")
                            .font(.caption)
                            .foregroundStyle(.brandPrimary)
                            .onTapGesture { onTagTap(tag) }
                    }
                }
            }
        }
        .padding(.horizontal, .spaceLG)
        .padding(.bottom, .spaceSM)
    }

    @ViewBuilder
    private var mediaSection: some View {
        MediaGrid(imageURLs: imageURLs)
            .padding(.horizontal, .spaceLG)
            .padding(.bottom, .spaceSM)
    }

    private var actionBar: some View {
        HStack(spacing: 0) {
            actionButton(
                icon: isLiked ? "heart.fill" : "heart",
                count: likeCount,
                color: isLiked ? .brandPrimary : .textTertiary,
                action: onLike
            )
            actionButton(
                icon: "bubble.left",
                count: commentCount,
                color: .textTertiary,
                action: onComment
            )
            actionButton(
                icon: "square.and.arrow.up",
                count: nil,
                color: .textTertiary,
                action: onShare
            )
            Spacer()
            Button(action: onMore) {
                Image(systemName: "ellipsis")
                    .font(.system(size: IconSize.md.rawValue))
                    .foregroundStyle(.textTertiary)
            }
            .frame(width: 44, height: 44)
        }
        .padding(.horizontal, .spaceLG)
        .frame(height: 44)
    }

    private func actionButton(icon: String, count: Int?, color: Color, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: .spaceXXS) {
                Image(systemName: icon)
                    .font(.system(size: IconSize.md.rawValue))
                    .foregroundStyle(color)
                if let count {
                    Text("\(count)")
                        .font(.caption)
                        .foregroundStyle(.textTertiary)
                }
            }
        }
        .frame(height: 44)
    }
}

struct MediaGrid: View {
    let imageURLs: [URL]

    var body: some View {
        let cols = gridColumns
        LazyVGrid(columns: cols, spacing: 2) {
            ForEach(imageURLs, id: \.self) { url in
                AsyncImage(url: url) { phase in
                    switch phase {
                    case .success(let image):
                        image.resizable().scaledToFill()
                    default:
                        Rectangle().fill(Color.bgTertiary)
                    }
                }
                .aspectRatio(1, contentMode: .fit)
                .clipShape(RoundedRectangle(cornerRadius: .radiusMD))
            }
        }
    }

    private var gridColumns: [GridItem] {
        switch imageURLs.count {
        case 1: return [GridItem(.flexible())]
        case 2: return [GridItem(.flexible()), GridItem(.flexible())]
        default: return [GridItem(.flexible()), GridItem(.flexible()), GridItem(.flexible())]
        }
    }
}

File: android/app/src/main/java/com/thirtysevendegrees/app/designsystem/molecules/PostCard.kt
package com.thirtysevendegrees.app.designsystem.molecules

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.thirtysevendegrees.app.designsystem.atoms.Avatar
import com.thirtysevendegrees.app.designsystem.tokens.*

@Composable
fun PostCard(
    avatarUrl: String?,
    username: String,
    handle: String,
    timeAgo: String,
    content: String,
    tags: List<String>,
    imageUrls: List<String>,
    likeCount: Int,
    commentCount: Int,
    isLiked: Boolean,
    onLike: () -> Unit,
    onComment: () -> Unit,
    onShare: () -> Unit,
    onMore: () -> Unit,
    onAvatarTap: () -> Unit,
    onTagTap: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    Column(modifier = modifier.background(AppColors.bgPrimaryLight)) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = Spacing.LG, vertical = Spacing.SM),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Avatar(
                url = avatarUrl, size = AvatarSize.LG, placeholder = username,
                modifier = Modifier.clickable(onClick = onAvatarTap)
            )
            Spacer(modifier = Modifier.width(Spacing.SM))
            Column {
                Text(username, style = MaterialTheme.typography.headlineMedium, maxLines = 1)
                Row(horizontalArrangement = Arrangement.spacedBy(Spacing.XS)) {
                    Text("@$handle", style = MaterialTheme.typography.labelSmall, color = AppColors.textTertiaryLight)
                    Text("·", color = AppColors.textTertiaryLight)
                    Text(timeAgo, style = MaterialTheme.typography.labelSmall, color = AppColors.textTertiaryLight)
                }
            }
        }

        if (content.isNotEmpty()) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = Spacing.LG)
                    .padding(bottom = Spacing.SM)
            ) {
                Text(
                    content,
                    style = MaterialTheme.typography.bodyMedium,
                    maxLines = 6,
                    overflow = TextOverflow.Ellipsis,
                    color = AppColors.textPrimaryLight
                )
                if (tags.isNotEmpty()) {
                    Row(horizontalArrangement = Arrangement.spacedBy(Spacing.XS)) {
                        tags.forEach { tag ->
                            Text(
                                "#$tag",
                                style = MaterialTheme.typography.labelSmall,
                                color = AppColors.brandPrimary,
                                modifier = Modifier.clickable { onTagTap(tag) }
                            )
                        }
                    }
                }
            }
        }

        if (imageUrls.isNotEmpty()) {
            MediaGrid(
                imageUrls = imageUrls,
                modifier = Modifier.padding(horizontal = Spacing.LG, vertical = Spacing.SM)
            )
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(44.dp)
                .padding(horizontal = Spacing.LG),
            verticalAlignment = Alignment.CenterVertically
        ) {
            ActionButton(
                icon = if (isLiked) Icons.Default.Favorite else Icons.Default.FavoriteBorder,
                count = likeCount,
                tint = if (isLiked) AppColors.brandPrimary else AppColors.textTertiaryLight,
                onClick = onLike
            )
            ActionButton(icon = Icons.Default.ChatBubbleOutline, count = commentCount, tint = AppColors.textTertiaryLight, onClick = onComment)
            ActionButton(icon = Icons.Default.Share, count = null, tint = AppColors.textTertiaryLight, onClick = onShare)
            Spacer(modifier = Modifier.weight(1f))
            IconButton(onClick = onMore, modifier = Modifier.size(44.dp)) {
                Icon(Icons.Default.MoreHoriz, contentDescription = null, tint = AppColors.textTertiaryLight)
            }
        }
    }
}

@Composable
private fun ActionButton(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    count: Int?,
    tint: Color,
    onClick: () -> Unit
) {
    Row(
        modifier = Modifier
            .height(44.dp)
            .clickable(onClick = onClick)
            .padding(horizontal = Spacing.SM),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(IconSize.MD))
        if (count != null) {
            Spacer(modifier = Modifier.width(Spacing.XXS))
            Text("$count", style = MaterialTheme.typography.labelSmall, color = AppColors.textTertiaryLight)
        }
    }
}

@Composable
fun MediaGrid(imageUrls: List<String>, modifier: Modifier = Modifier) {
    val columns = when (imageUrls.size) {
        1 -> 1
        2 -> 2
        else -> 3
    }
    LazyVerticalGrid(
        columns = GridCells.Fixed(columns),
        horizontalArrangement = Arrangement.spacedBy(2.dp),
        verticalArrangement = Arrangement.spacedBy(2.dp),
        modifier = modifier.heightIn(max = 300.dp),
        userScrollEnabled = false
    ) {
        items(imageUrls) { url ->
            AsyncImage(
                model = url,
                contentDescription = null,
                modifier = Modifier
                    .aspectRatio(1f)
                    .clip(CornerRadius.MD),
                contentScale = ContentScale.Crop
            )
        }
    }
}
